### PR TITLE
Implement a function to defer rendering via promise chaining

### PR DESCRIPTION
**Pull request by danmcclain**
Service side of
https://github.com/ember-fastboot/ember-fastboot-server/pull/41

### CODE REVIEW

File: tests/fixtures/async-content/app/components/async-content.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    this._super(...arguments);
+
+    let deferred = Ember.RSVP.defer();
+
+    Ember.run.later(() => {
+      if (!this.get('isDestroyed')) {
+        this.set('setLater', 'foo');
+      }
+      deferred.resolve()
+    }, 100);
```
Comment (danmcclain) [inline tests/fixtures/async-content/app/components/async-content.js]: This is enough of a delay to escape the initial `visit` call

**Comment by danmcclain** [comment]
@tomdale @stefanpenner Updated per comments on #180

File: app/services/fastboot.js [hunk 2]
```diff
@@ -46,5 +46,8 @@ export default Ember.Service.extend({
 
   isFastBoot: computed(function() {
     return typeof FastBoot !== 'undefined';
-  })
+  }),
+  deferRendering(promise) {
```
Comment (stefanpenner) [inline app/services/fastboot.js]: don't we more typically have hooks, the when invoked do.

new Promise(resolve => resolve(cb());

Passing promises around like this doesn't really align with other hooks, like model routes hooks, or adapter hooks.

File: tests/acceptance/async-content-test.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  var app;
+
+  before(function() {
+
+    app = new AddonTestApp();
+
+    return app.create('async-content')
+      .then(function() {
+        debugger;
```
Comment (stefanpenner) [inline tests/acceptance/async-content-test.js]: ^

File: tests/fixtures/async-content/app/components/async-content.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  },
+
+  willDestroy() {
+    let runLater = this.get('runLater');
+    let deferred = this.get('deferred');
+
+    if (runLater) {
+      Ember.run.cancel(runLater);
+      deferred.promise.catch(function() {
+        console.log('swallow reject');
```
Comment (stefanpenner) [inline tests/fixtures/async-content/app/components/async-content.js]: we should let it throw, so it hits `Ember.onerror`

File: tests/fixtures/async-content/app/components/async-content.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      this.set('deferred', null);
+    }, 100);
+
+    this.runLater =  runLater;
+    this.deferred = deferred;
+
+    this.get('fastboot').deferRendering(deferred.promise);
+  },
+
+  willDestroy() {
```
Comment (stefanpenner) [inline tests/fixtures/async-content/app/components/async-content.js]: `destroy`

File: app/services/fastboot.js (hunk 2 above)
Comment (danmcclain) [inline app/services/fastboot.js]: If we were to add a hook, where would we introduce it?

This is to handle the case that you have asynchrony outside of the normal hooks. as such, you may be doing this work inside a component.

Or are you suggesting that we make a new promise from the promise being passed into `deferRendering`

**Pull request merged**